### PR TITLE
Allow resizing Mir-on-X windows

### PR DESCRIPTION
**Pull request by wmww**
This should work. Mostly. You can still make the background fail by resizing constantly for too long and there's some glitches when there are multiple windows, but I think good enough to land.

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
This seems to work on GNOME/Wayland, but not on GNOME/Xorg.

**Comment by wmww** [comment]
What do you mean by "work"? Is it glitchy? Are you unable to resize at all? If the latter, can you maximize?

**Comment by AlanGriffiths** [comment]
> What do you mean by "work"? Is it glitchy? Are you unable to resize at all? If the latter, can you maximize?

I get a resize cursor, but dragging doesn't change window size.

Maximize/unmaximize works

**Comment by AlanGriffiths** [comment]
Oh! I got something like #1900 on GNOME/Wayland without DisplayLink.

**Comment by wmww** [comment]
Fixes #1314

**Comment by wmww** [comment]
There's some flickering and there seems to be some buffering (with lots of size changes it gets behind and then needs to "catch up", but for me it's usable and doesn't crash. I haven't tried it on GNOME yet.

**Comment by AlanGriffiths** [comment]
OK, the problems I was seeing are my separate issues with Xwayland on Displaylink. Could we avoid queueing the repaints by not triggering a redraw until the mouse is released? This doesn't look good:

https://user-images.githubusercontent.com/9048879/114004365-7e63a100-9856-11eb-87ac-ae2d73a21e39.mp4

**Comment by wmww** [comment]
Hm, I'll think about that. I'd rather do it by figuring out where updates are being buffered and short-circuit that (so it doesn't have to process older states when newer ones are available). If that doesn't end up being practical not sending updates until mouse up is a reasonable fallback.

**Comment by wmww** [comment]
Perhaps that's a little better? It still doesn't apply the changes instantly so there's some jank, but it now skips updates when they come faster than they're processed.

File: src/server/scene/mediating_display_changer.h [hunk 1]
```diff
@@ -122,6 +122,12 @@ class MediatingDisplayChanger : public frontend::DisplayChanger,
     std::weak_ptr<scene::Session> currently_previewing_session;
     struct SessionObserver;
     std::unique_ptr<SessionObserver> const session_observer;
+
+    /// Mutex protecting pending_configuration
+    std::mutex pending_configuration_mutex;
+    /// See configure_for_hardware_change(). The config that will be applied by a currently in-flight server action.
+    /// If null, there is no config to apply or hardware config server action queued.
+    std::shared_ptr<graphics::DisplayConfiguration> pending_configuration;
```
Comment (AlanGriffiths) [inline src/server/scene/mediating_display_changer.h:130]: I was about to suggest `atomic<shared_ptr<...>>`, but that's C++20. One for another year...

**Review by AlanGriffiths — APPROVED** [review]
Works well enough to land

bors r+

**Pull request merged**